s3/replication/common/src/s3replicationcommon/s3_range_read.py

```diff
@@ -0,0 +1,134 @@
+    def get_state(self):
+        """Returns current request state."""
+        return self._state
+
+    def get_execution_time(self):
+        """Return total time for GET Object operation."""
+        return self._timer.elapsed_time_ms()
+
+    def get_etag(self):
+        """Returns ETag for object."""
+        return self._response_headers["ETag"].strip("\"")
+
+    def get_content_range(self):
+        """Get content range in bytes."""
+        return self._response_headers["Content-Range"]
+
+    # yields data chunk for given size
+    async def fetch(self):
+        request_uri = AWSV4Signer.fmt_s3_request_uri(
+            self._bucket_name, self._object_name)
+
+        query_params = ""
+        body = ""
+        obj_range = self._object_range
+
+        headers = AWSV4Signer(
+            self._session.endpoint,
+            self._session.service_name,
+            self._session.region,
+            self._session.access_key,
+            self._session.secret_key).prepare_signed_header(
+            'GET',
+            request_uri,
+            query_params,
+            body,
+            obj_range)
+
+        if (headers['Authorization'] is None):
+            self._logger.error(fmt_reqid_log(self._request_id) +
+                               "Failed to generate v4 signature")
+            sys.exit(-1)
+
+        self._logger.info(fmt_reqid_log(self._request_id) +
+                          'GET on {}'.format(
+                              self._session.endpoint + request_uri))
+        self._logger.debug(fmt_reqid_log(self._request_id) +
+                           "GET with headers {}".format(headers))
+        self._timer.start()
+        try:
+            async with self._session.get_client_session().get(
+                    self._session.endpoint + request_uri,
+                    headers=headers) as resp:
+
+                self._logger.info(fmt_reqid_log(self._request_id)
+                                  + 'GET response received with'
+                                  + ' status code: {}'.format(resp.status))
+                self._logger.info('Response url {}'.format(
+                    self._session.endpoint + request_uri))
+
+                if resp.status == 206:
+                    self._response_headers = resp.headers
+                    self._logger.info('Response headers {}'.format(
+                        self._response_headers))
+                else:
+                    self._state = S3RequestState.FAILED
+                    self._logger.error(
+                        fmt_reqid_log(self._request_id) +
+                        'GET Object failed with http status: {}'.
+                        format(resp.status) +
+                        '\nError Response: Invalid object range')
+                    return
+
+                while True:
+                    resp_data = await resp.content.read()
+                    if not resp_data:
+                        break
+                    yield resp_data
```

Comment:
do we need a generator here? we do not read in chunks, it is read() not read(length)

Reply:
yes. done